### ISSUE TITLE
@types/reactstrap doesn't allow custom props

### ISSUE THREAD

**Issue by mikfoo** [DefinitelyTyped/DefinitelyTyped#20461]
- [x] I tried using the `@types/reactstrap` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @alihammad @mfal @danilobjr @fabiopaiva

Basically I'm trying to use Reactstrap together with React Router 4, and have a line like this:
```
<NavbarBrand tag={Link} to="/">
```
Since the `<Link>` provided by React Router requires the prop `to`, this fails. I get the following error:
```
error TS2339: Property 'to' does not exist on type 'IntrinsicAttributes & Props & { children?: ReactNode; }'.
```

**Comment by fabiopaiva** [comment]
The correct prop is `href`, this component is not related to React Router.
To use ReactRouter you can do something like this:
```
<NavbarBrand tag="div">
  <Link to="/">
    <img />
  </Link>
</NavbarBrand>
```

**Comment by mikfoo** [comment]
Yes, but for `tag` I'm also able to pass in a component/function, in this case the `Link` that is provided by React Router (it could be any other component though). I think it's called stateless components? In any case...

`Link` from React Router accepts `to`, `exact`, and so on.

When doing `<NavbarBrand tag={Link} to="/foo">` the NavBar will create the `Link` component, add its own values, and then and add the extra props.

See code: https://github.com/reactstrap/reactstrap/blob/master/src/NavbarBrand.js

This is also true for many of the other components in Reactstrap, including `NavbarLink`, `DropdownItem`, `Collapse`, and so on.

My temporary (and not really elegant) fix for this is just doing:

```jsx
const NavbarBrandX: React.StatelessComponent<any> = NavbarBrand
const NavLinkX: React.StatelessComponent<any> = NavLink
const DropdownItemX: React.StatelessComponent<any> = DropdownItem
const CollapseX: React.StatelessComponent<any> = Collapse
```

**Comment by fabiopaiva** [comment]
I see what you mean, I don't know how is the correct way to declare that, sorry.

**Comment by mikfoo** [comment]
I've toyed a little around, and I believe doing something like this should work:

```typescript
import { CSSModule } from '../index';

interface Props extends React.HTMLProps<HTMLAnchorElement> {
  tag?: React.ReactType
  className?: string;
  cssModule?: CSSModule;
  [key: string]: any // <-- This line is new
}

declare var NavbarBrand: React.StatelessComponent<Props>;
export default NavbarBrand;
```
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/reactstrap/lib/NavbarBrand.d.ts

Of course all the other types needs to be updated as well.

**Comment by corydeppen** [comment]
Do we have consensus yet on the need and benefit of adding a dynamic property to a few of the link-related components? There have been a few other similar discussions in the past (e.g. #17123) and I'm guessing there are many folks maintaining separate definition overrides to handle these situations. It would be nice to support these cases in the official definitions. Should be a quick change if we can identify the components that need updating.

/cc @alihammad @mfal @danilobjr

**Comment by mikfoo** [comment]
I think any Reactstrap component that accepts a `tag` prop must also support extra optional props as well, this is not only relevant for `Button` with `Link`.

If it doesn't support extra props, then it makes the `tag` completely useless in most cases, and makes it difficult to work with Reacstrap and TS.

**Comment by fabiopaiva** [comment]
I think it's not that simple. Let suppose you have a dynamic variable, `myTag`, which could be different tags, `div`, `a`, Link or whatever... You only could detect that on runtime, doesn't fit at all

**Comment by mikfoo** [comment]
It's true that it would be checked on runtime, but at least you would be able to use the `tag` prop as well as those extra props the tag might need, vs. not being able to use it at all (unless you override the definition(s) locally).

I've seen quite a few other typings use `[key: string]: any` for when they support passing in custom components.

Personally I think this is the best solution for now, even if it's not the most optimal one. At least we get to keep some type safety this way, while allowing any other optional prop as well.

Is it possible to "extend" the typings locally? I tried messing around but I had issues with "module can't be redclared" and such.

**Comment by fabiopaiva** [comment]
Maybe this works:
```
```
import * as reactstrap from 'reactstrap';

declare module 'reactstrap' {
  interface Props extends React.HTMLProps<HTMLAnchorElement> {
    tag?: React.ReactType;
    className?: string;
    cssModule?: reactstrap.CSSModule;
    to?: string;
  }

  var NavbarBrand: React.StatelessComponent<Props>;
}
```

**Comment by mikfoo** [comment]
I tried something similar earlier, but I get the error:

```
error TS2323: Cannot redeclare exported variable 'NavbarBrand'.
```

**Comment by fabiopaiva** [comment]
To me it works, but after restarting my `react-scripts-ts` I get the same error. Sorry, it's more weird than I expected

**Comment by AndyCJ** [comment]
This is also a problem on the Input component.

reactstrap itself allows [arbitrary props](https://github.com/reactstrap/reactstrap/blob/master/src/Input.js#L44) to be [passed through](https://github.com/reactstrap/reactstrap/blob/master/src/Input.js#L102). 

But the typings only expose the props on [`React.InputHTMLAttributes<HTMLInputElement>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/reactstrap/v4/lib/Input.d.ts#L32) combined with a few [reactstrap specific props](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/reactstrap/v4/lib/Input.d.ts#L36).

In my particular case I just want to use the [rows attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#attr-rows) since my Input has a type of textarea.

I've found I can work around this by using the spread operator:

````typescript
import * as ReactStrap from "reactstrap";

const rows = 5;

const attributesHack = {
    rows
};

<ReactStrap.Input type="textarea" {...attributesHack} />
````

As mikfoo mentioned above, adding `[key: string]: any` would mean the typings accurately reflect what the reactstrap code is doing/allowing.

**Comment by FaithForHumans** [comment]
Fixed by #26094!

**Comment by FaithForHumans** [comment]
There's tests to prove the functionality is there. There's two ways to achieve any generic property, [the first way is shown starting here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9c32df15f4fa09780528d0d28642337937d95052/types/reactstrap/reactstrap-tests.tsx#L3756) and [the second way is shown starting here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9c32df15f4fa09780528d0d28642337937d95052/types/reactstrap/reactstrap-tests.tsx#L3925).

I haven't tested either of the following code, so there might be a typo or two, but they both should work.

So that means you can either do something like:

    import { NavLink as RRNavLink } from 'react-router-dom';
    import { Nav, NavItem } from 'reactstrap';
    // Note relative import to allow any property
    import { default as NavLink } from 'reactstrap/lib/NavLink'; 
    
    const NavigationAuth = () =>
      <Nav className='ml-auto' navbar={true}>
        <NavItem>
          <NavLink to={routes.HOME} tag={RRNavLink} >Home</NavLink>
        </NavItem>
      </Nav>

or something like:

    import { NavLink as RRNavLink } from 'react-router-dom';
    import { Nav, NavItem, NavLink } from 'reactstrap';

    // extend to just get the property you want.
    class ReactstrapNavLinkRouter extends NavLink< {to?: string} > {}; 
    
    const NavigationAuth = () =>
      <Nav className='ml-auto' navbar={true}>
        <NavItem>
          <ReactstrapNavLinkRouter to={routes.HOME} tag={RRNavLink} >Home</NavLink>
        </NavItem>
      </Nav>

whichever suits your preferences more.

**Comment by FaithForHumans** [comment]
It has to deal with how the components are exported. In general, the component props are structured such that they follow the pattern of:

    interface ComponentProps<T={}> = T & *relevant HTML Props* & *reactstrap specific props*;

In the reactstrap/lib/ directory, the component is exported following the pattern:

    declare class Component<T = {[key: string]: any} > extends React.Component<ComponentProps<T>> {}
    export default Component

While in the index.d.ts file, they're exported following the pattern of:

    import { Component as Component_ } from 'lib/Component';
    export class Component<T = {}> extends Component_<T> {}

So it all comes down to what that extended interface is defaulting to. One of the other authors didn't want the `{[key: string]: any}` to be the default export which is why the index.d.ts exports it as 'nothing'. The `{[key: string]: any}` is there in the lib/ directory for ease of use.

**Comment by Psvensso** [comment]
@FaithForHumans Thank you for replying! 
I read it the other way around first so that `import { Modal } from "reactstrap";` would require a generic. 

However the functionality you describe seems broken in TS 2.9+

We are using TypeScript 3.*.* in my team and i cannot pinpoint the version when this feature stopped working but my guess is that its related to the TSX Generics around TS 2.9.
I know for a fact that rollbacking TS to 2.8 does give a compile warning on 

```
import { Modal } from "reactstrap";
<Modal badOptionalType={true} /> //badOptionalType does not excist on.....
```

However on TS 2.9 > (or as i said atleast confirmed on 3.0 >) you have to specify the generi T on the JSX element for the strict check to work. 

```
import { Modal } from "reactstrap";

<Modal badOptionalType={true} /> //This is ok in 3.0! Even if type is Modal<T = {}>

<Modal<{}> badOptionalType={true} /> //Not ok in 3.0+ (badOptionalType does not excist on.....)
```

I can get any further on this "bug" and just wanted to share my findings here, perhaps someone else knows whats going on.

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
